src/NServiceBus.Core.Tests/DelayedDelivery/TimeoutManager/ExpiredTimeoutsPollerTests.cs

```diff
@@ -0,0 +1,105 @@
+
+        [SetUp]
+        public void SetUp()
+        {
+            breaker = new FakeBreaker();
+            dispatcher = new RecordingFakeDispatcher();
+            timeouts = new InMemoryTimeoutPersister();
+            poller = new ExpiredTimeoutsPoller(timeouts, dispatcher, "test", breaker, TimeSpan.Zero, TimeSpan.Zero);
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            poller.Dispose();
+        }
+
+        [Test]
+        public async void Sends_no_messages_when_no_timeouts_registered()
+        {
+            await SpinOnce();
+            CollectionAssert.IsEmpty(dispatcher.DispatchedMessages);
+        }
+
+        /// <summary>
+        /// Refers to<see cref= "http://github.com/Particular/NServiceBus/issues/3300" />
+        /// </summary >
+        [Test]
+        public async void Updates_next_retrieval_time_when_timeout_registered_in_the_middle_and_dispatches_the_message()
+        {
+            await SpinOnce();
+
+            var nextRetrieval = poller.NextRetrieval;
+            var newTimeout = nextRetrieval.Subtract(HalfOfDefaultInMemoryPersisterSleep);
+
+            RegisterNewTimeout(newTimeout);
+            Assert.AreEqual(nextRetrieval, poller.NextRetrieval);
+
+            await SpinOnce();
+
+            Assert.AreEqual(nextRetrieval, poller.NextRetrieval);
+            Assert.AreEqual(1, dispatcher.DispatchedMessages.Count);
+        }
+
+        /// <summary>
+        /// Refers to<see cref= "http://github.com/Particular/NServiceBus/issues/3300" />
+        /// </summary >
+        [Test]
+        public async void Keeps_next_retrieval_equal_to_the_registered_timeout_even_when_persister_returns_more_timeouts_and_dispatches_both_timeouts()
+        {
+            var nextRetrieval = poller.NextRetrieval;
+            var timeout1 = nextRetrieval.Subtract(HalfOfDefaultInMemoryPersisterSleep);
+            // ReSharper disable once PossibleLossOfFraction
+            var timeout2 = timeout1.Add(TimeSpan.FromMilliseconds(HalfOfDefaultInMemoryPersisterSleep.Milliseconds/2));
+
+            RegisterNewTimeout(timeout1);
+            RegisterNewTimeout(timeout2, false);
+
+            await SpinOnce();
+
+            Assert.AreEqual(timeout1, poller.NextRetrieval);
+            Assert.AreEqual(2, dispatcher.DispatchedMessages.Count);
+        }
+
+        private void RegisterNewTimeout(DateTime newTimeout, bool withNotification = true)
+        {
+            timeouts.Add(new TimeoutData { Time = newTimeout }, null);
+            if (withNotification)
+            {
+                poller.NewTimeoutRegistered(newTimeout);
+            }
+        }
+
+        private Task SpinOnce()
+        {
+            return (Task)typeof(ExpiredTimeoutsPoller).
+                InvokeMember("SpinOnce", BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.InvokeMethod, null, poller, new object[0]);
```

Comment:
Yes, it can. I wanted to keep the change minimal. I didn't want to introduce a new interface or marking that internal. If making sth internal only for tests purposes is ok, I'll apply this.